html/mycakephpapp/src/Controller/AuctionController.php

```diff
@@ -212,5 +218,70 @@ public function home2() {
         ])->toArray();
         $this->set(compact('biditems'));
     }
+
+    // 落札後ページの表示
+    public function end($id) {
+
+        $rating = $this->Ratings->newEntity();
+        // $idのBiditemを取得
+        $biditem = $this->Biditems->get($id, [
+            'contain' => ['Users', 'Bidinfo', 'Bidinfo.Users']
+        ]);
+
+        $bidinfo = $this->Bidinfo->find('all', [
+            'contain' => ['Users', 'Biditems', 'Biditems.Users']
+        ])->where(['biditem_id' => $id])->first();
+
+        $messages = $this->Bidmessages->find('all', [
+            'contain' => ['Users']
+        ])->where(['bidinfo_id' => $bidinfo->id])->toArray();
+        
+        //Ratingテーブルにすでに評価がされているかをチェック
+        $record_check = $this->Ratings->find('all')
+            ->where(['bidinfo_id' => $biditem->bidinfo->id])->count();
+        //評価ががすでにある場合はViewにその評価内容を渡す
+        if ($record_check) {
+            $rating = $this->Ratings->find('all')
+            ->where(['bidinfo_id' => $biditem->bidinfo->id])->first();
+        }
+
+        if ($this->request->isPost() && $this->request->getData('Bidinfo')) {
+            // 送信されたフォームで$bidinfoを更新
+            $bidinfo = $this->Bidinfo->patchEntity($bidinfo, $this->request->getData('Bidinfo'));
+            // Bidinfoを保存
+            if ($this->Bidinfo->save($bidinfo)) {
+                //結果を再度取得
+                $bidinfo = $this->Bidinfo->find('all', [
+                    'contain' => ['Users', 'Biditems', 'Biditems.Users']
+                ])->where(['biditem_id' => $id])->first();
+                // 成功時のメッセージ
+                $this->Flash->success(__('保存しました。'));
+            } else {
+                //結果を再度取得
+                $bidinfo = $this->Bidinfo->find('all', [
+                    'contain' => ['Users', 'Biditems', 'Biditems.Users']
+                ])->where(['biditem_id' => $id])->first();
+                $this->Flash->error(__('保存に失敗しました。もう一度入力下さい。'));
+            }
+        } elseif ($this->request->isPost() && $this->request->getData('Ratings')) {
+
+            // 送信されたフォームで$ratingsを更新
+            $rating = $this->Ratings->patchEntity($rating, $this->request->getData('Ratings'));
+            // Ratingsを保存
+            if ($this->Ratings->save($rating)) {
+                //結果を再度取得
+                $rating = $this->Ratings->find('all')
+                ->where(['bidinfo_id' => $biditem->bidinfo->id])->first();
+                // 成功時のメッセージ
+                $this->Flash->success(__('保存しました。'));
+            } else {
+                //結果を再度取得
+                $rating = $this->Ratings->find('all')
+                ->where(['bidinfo_id' => $biditem->bidinfo->id])->first();
+                $this->Flash->error(__('保存に失敗しました。もう一度入力下さい。'));
+            }
```

Comment:
(動作検証していません)再取得せずに以下のように書くこともできそうです。
```suggestion
            // 送信されたフォームで$ratingsを更新
            $rating_new = $this->Ratings->patchEntity($rating, $this->request->getData('Ratings'));
            // Ratingsを保存
            if ($this->Ratings->save($rating_new)) {
                $rating = $rating_new;
                // 成功時のメッセージ
                $this->Flash->success(__('保存しました。'));
            } else {
                $this->Flash->error(__('保存に失敗しました。もう一度入力下さい。'));
            }
```

Reply:
DBへのアクセスを減らすことができますね。ご指摘いただきました内容を理解した上で修正しました。